staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go

```diff
@@ -608,211 +621,50 @@ See http://k8s.io/docs/reference/using-api/api-concepts/#conflicts`, err)
-			FieldSelector: p.fieldSelector,
-		})
-	if err != nil {
-		return err
-	}
-
-	objs, err := meta.ExtractList(objList)
+// MarkNamespaceVisited keeps track of UIDs of the applied
+// objects. Used for pruning.
+func (o *ApplyOptions) MarkObjectVisited(info *resource.Info) error {
+	metadata, err := meta.Accessor(info.Object)
 	if err != nil {
 		return err
 	}
-
-	for _, obj := range objs {
-		metadata, err := meta.Accessor(obj)
-		if err != nil {
-			return err
-		}
-		annots := metadata.GetAnnotations()
-		if _, ok := annots[corev1.LastAppliedConfigAnnotation]; !ok {
-			// don't prune resources not created with apply
-			continue
-		}
-		uid := metadata.GetUID()
-		if p.visitedUids.Has(string(uid)) {
-			continue
-		}
-		name := metadata.GetName()
-		if !p.dryRun {
-			if err := p.delete(namespace, name, mapping); err != nil {
-				return err
-			}
-		}
-
-		printer, err := p.toPrinter("pruned")
-		if err != nil {
-			return err
-		}
-		printer.PrintObj(obj, p.out)
-	}
+	o.visitedUids.Insert(string(metadata.GetUID()))
 	return nil
 }
 
-func (p *pruner) delete(namespace, name string, mapping *meta.RESTMapping) error {
-	return runDelete(namespace, name, mapping, p.dynamicClient, p.cascade, p.gracePeriod, p.serverDryRun)
-}
-
-func runDelete(namespace, name string, mapping *meta.RESTMapping, c dynamic.Interface, cascade bool, gracePeriod int, serverDryRun bool) error {
-	options := &metav1.DeleteOptions{}
-	if gracePeriod >= 0 {
-		options = metav1.NewDeleteOptions(int64(gracePeriod))
-	}
-	if serverDryRun {
-		options.DryRun = []string{metav1.DryRunAll}
-	}
-	policy := metav1.DeletePropagationForeground
-	if !cascade {
-		policy = metav1.DeletePropagationOrphan
-	}
-	options.PropagationPolicy = &policy
-	return c.Resource(mapping.Resource).Namespace(namespace).Delete(name, options)
+func (o *ApplyOptions) SetPreProcessorFn(pre func(*ApplyOptions) error) {
+	o.preprocessFn = pre
 }
 
-func (p *Patcher) delete(namespace, name string) error {
-	return runDelete(namespace, name, p.Mapping, p.DynamicClient, p.Cascade, p.GracePeriod, p.ServerDryRun)
+func (o *ApplyOptions) SetPostProcessorFn(post func(*ApplyOptions) error) {
+	o.preprocessFn = post
 }
 
-// Patcher defines options to patch OpenAPI objects.
-type Patcher struct {
-	Mapping       *meta.RESTMapping
-	Helper        *resource.Helper
-	DynamicClient dynamic.Interface
+// Prints applied objects (depending on print options) and prunes
+// (if prune is set). Used as the standard apply postprocessor.
+func PrintAndPrune(o *ApplyOptions) error {
 
-	Overwrite bool
-	BackOff   clockwork.Clock
-
-	Force        bool
-	Cascade      bool
-	Timeout      time.Duration
-	GracePeriod  int
-	ServerDryRun bool
-
-	// If set, forces the patch against a specific resourceVersion
-	ResourceVersion *string
+	if err := o.printObjects(); err != nil {
+		return err
+	}
 
-	// Number of retries to make if the patch fails with conflict
-	Retries int
+	if o.Prune {
+		p := newPruner(o)
+		return p.pruneAll(o)
```

Comment:
This can be cleaned up, by adding a couple fields that are stored in `ApplyOptions`, but which aren't stored in the `pruner`. I hadn't added those fields yet, because I wanted to keep the code changes to a minimum. That cleanup is scheduled for a future PR, but I can do it now if you like.